openwisp_controller/config/api/serializers.py

```diff
@@ -0,0 +1,222 @@
+import ast
+import collections
+import json
+
+from django.db import transaction
+from django.utils.translation import ugettext_lazy as _
+from rest_framework import serializers
+from swapper import load_model
+
+from openwisp_users.api.mixins import FilterSerializerByOrgManaged
+from openwisp_utils.api.serializers import ValidatedModelSerializer
+
+Template = load_model('config', 'Template')
+Vpn = load_model('config', 'Vpn')
+Device = load_model('config', 'Device')
+Config = load_model('config', 'Config')
+TemplateTag = load_model('config', 'TemplateTag')
+Organization = load_model('openwisp_users', 'Organization')
+
+
+class BaseMeta:
+    read_only_fields = ['created', 'modified']
+
+
+class BaseSerializer(FilterSerializerByOrgManaged, ValidatedModelSerializer):
+    pass
+
+
+class TemplateSerializer(BaseSerializer):
+    config = serializers.JSONField(initial={})
+    tags = serializers.StringRelatedField(many=True, read_only=True)
+    default_values = serializers.JSONField(required=False, initial={})
+
+    class Meta(BaseMeta):
+        model = Template
+        fields = [
+            'id',
+            'name',
+            'tags',
+            'organization',
+            'type',
+            'backend',
+            'default',
+            'required',
+            'default_values',
+            'config',
+            'created',
+            'modified',
+        ]
+
+
+class VpnSerializer(BaseSerializer):
+    config = serializers.JSONField(initial={})
+
+    class Meta(BaseMeta):
+        model = Vpn
+        fields = [
+            'id',
+            'name',
+            'host',
+            'organization',
+            'key',
+            'ca',
+            'cert',
+            'backend',
+            'notes',
+            'dh',
+            'config',
+        ]
+
+
+class BaseJsonField(serializers.Field):
+    def to_internal_value(self, data):
+        if type(data) is str:
+            if len(data) < 2:
+                data = '{}'
+            return ast.literal_eval(data)
+        else:
+            return data
+
+
+class JsonConfigField(BaseJsonField):
+    def to_representation(self, value):
+        return json.loads(json.dumps(value))
+
+
+class JsonContextField(BaseJsonField):
+    def to_representation(self, value):
+        return dict(value)
```

Comment:
> Ok you can leave this thing of the JSON fields as last thing, please can you point out which is the exact piece of code which influences this behavior?

@nemesisdesign 
Usually in other cases where the default field is JSON is handled with `config = serializers.JSONField()`(Please have a look at other serializers), But in these case things were not working as expected so had to overwrite the `to_representation` and `to_internal_value`. You can experiment with this pieces of code.